tests/pass/float.rs

```diff
@@ -148,155 +212,266 @@ fn basic() {
+        let izero: $ity = 0;
+        #[allow(unused_comparisons)]
+        let isigned = (0 as $ity).saturating_sub(1) < 0;
+
+        #[allow(overflowing_literals)]
+        let imin_f: $fty = $imin_f;
+        #[allow(overflowing_literals)]
+        let imax_f: $fty = $imax_f;
+
+        // If an integer can fit entirely in the mantissa (counting the hidden bit), every value
+        // can be represented exactly.
+        let all_ints_exact_rep = ibits <= <$fty>::SIGNIFICAND_BITS + 1;
+
+        // Skip unchecked cast if negative numbers are unrepresentable
+        let assert_ftoi_neg = if isigned { assert_ftoi } else { assert_ftoi_unrep };
+        // Skip unchecked cast when int min/max would be unrepresentable
+        let assert_ftoi_big = if all_ints_exact_rep { assert_ftoi } else { assert_ftoi_unrep };
+        let assert_bidir_big = if all_ints_exact_rep { assert_bidir } else { assert_bidir_unrep };
+
+        // Near zero representations
+        assert_bidir(0.0, 0, "zero");
+        assert_ftoi(-0.0, 0, "negative zero");
+        assert_ftoi_neg(-0.99999999999999999999999999, izero.saturating_sub(1), "near -1");
+        assert_ftoi(<$fty>::from_bits(0x1), 0, "min subnormal");
+        assert_ftoi(<$fty>::from_bits(0x1 | 1 << (fbits - 1)), 0, "min neg subnormal");
+
+        // Spot checks. Use `saturating_sub` to create negative integers so that unsigned
+        // integers stay at zero.
+        assert_ftoi(0.9, 0, "0.9");
+        assert_ftoi_neg(-0.9, 0, "-0.9");
+        assert_ftoi(1.1, 1, "1.1");
+        assert_ftoi_neg(-1.1, izero.saturating_sub(1), "-1.1");
+        assert_ftoi(1.9, 1, "1.9");
+        assert_ftoi_neg(-1.9, izero.saturating_sub(1), "-1.9");
+        assert_ftoi(5.0, 5, "5.0");
+        assert_ftoi_neg(-5.0, izero.saturating_sub(5), "-5.0");
+        assert_ftoi(5.9, 5, "5.0");
+        assert_ftoi_neg(-5.9, izero.saturating_sub(5), "-5.0");
+
+        // Exercise the middle of the integer's bit range. A single digit fits as long as the
+        // exponent can fit its offset, so cap at the largest shift representable by the
+        // exponent.
+        let half_i_max: $ity = 1 << (ibits / 2).min((1 << (<$fty>::EXPONENT_BITS - 1)) - 1);
+        let half_i_min = izero.saturating_sub(half_i_max);
+        assert_bidir(half_i_max as $fty, half_i_max, "half int max");
+        assert_bidir(half_i_min as $fty, half_i_min, "half int min");
+
+        // Integer limits
+        assert_bidir_big(imax_f, imax, "i max");
+        assert_bidir_big(imin_f, imin, "i min");
+        assert_ftoi_big(imax_f + 0.99, <$ity>::MAX, "slightly above i max");
+        assert_ftoi_neg(imin_f - 0.99, <$ity>::MIN, "slightly below i min");
+
+        if fbits == ibits {
+            // Signed integers have one fewer bit of range
+            let adj = if isigned { 1 } else { 0 };
+
+            let exp_overflow = 1 << (<$fty>::EXPONENT_BITS - 1 - adj);
+            let exp_max = (1 << <$fty>::EXPONENT_BITS - adj) - 1;
```

Comment:
I have absolutely no idea what this does

Reply:
I am not sure how to explain it concisely since I don't entirely understand it either, could use some suggestions. This comes from the "rounding loss" magic number tests:

```rust
test_both_cast::<f32, u32>((u32::MAX - 128) as f32, u32::MAX - 255); // rounding loss
assert_eq::<u32>((u32::MAX - 127) as f32 as u32, u32::MAX); // rounds up and then becomes unrepresentable

test_both_cast::<f64, u64>((u64::MAX - 1024) as f64, u64::MAX - 2047); // rounding loss
assert_eq::<u64>((u64::MAX - 1023) as f64 as u64, u64::MAX); // rounds up and then becomes unrepresentable
```

Basically it calculates the maximum value that will get rounded down to an integer less than `Int::MAX` (that exact value being `u32::MAX - 255` / `u64::MAX - 2047`), and checks the result. Then any number greater will get rounded up to `Int::MAX` or above.

I had to reverse engineer the pattern, `(exp_overflow, exp_max)` are my bad names for `(128, 255)` and `(1024, 2047)`. I renamed them to `(nearest_rep_offset, imax_offset)` but even those don't seem great. Signed numbers aren't tested in the original but the pattern holds if you shift the bits an extra position (as I do with `adj`).

Reply:
I think I added these tests. Basically I was looking for the largest integer that does round-trip properly with a float. For f32 and u32, that's `u32::MAX - 128` -- once you get any bigger, it becomes a float that saturates to `u32:MAX` (and causes UB with the unchecked casts).

I have no idea how to compute these numbers I found them experimentally.

Reply:
I tested this pattern with f16 and f128 before and I _think_ it holds true for those as well, but don't remember. https://github.com/rust-lang/rust/pull/126608 should hopefully be in pretty soon so I'll probably just wait for that then update this PR, to get the extra two cases to test the tests.

Reply:
If you want to stick to computing these numbers, then please add comments explaining what you are computing and how.

Is there a reason we can only have these tests when `fbits == ibits`? I couldn't be bothered to figure out the largest representable number for all combinations of types, but if there is a way to compute this automatically then that should be doable, right?

Reply:
Finally think I wrapped my head around this - the max integer that can be represented is a saturated mantissa (including the implicit bit), shifted to fully to the left in the integer. IOW, zero out the last F::EXPONENT_BITS of I::MAX if they have the same size, which is what subtracting 255 does (255 = 0b11111111, which is a f32's exponent mask). So `x = u32::MAX - 255`

Then the cutoff point between rounding to x and rounding to infinity will be midway between x and the integer's max, or adding back `x + 255/2`. `-255 + (255 / 2) = `-128`, hence that number. Signed integers don't use the top bit for their maximum, so they get one extra bit in their lower range compared to their same-sized unsigned integer. This is what I was representing with `adj`, without really understanding it.

I think this should be generalizable to anything where:

- `F::EXPONENT_BIAS >= I::BITS`, because it needs to be able to represent the magnitude of an integer with a learning `1`. Basically no f16.
- `I::BITS > (F::MANTISSA_BITS + 1)`, because if any integer bitpattern fits in the mantissa then this is pointless.

(This isn't so much an update for you, mostly just writing down my informal proof so I remember it :). [scratchpad](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d5ba0abff91d84238da3789ef0f90c6a) )